##################################
#04/02/2020

#VST Normalization --> Plotting PCAs

#Purpose:
#1) Normalize raw count matrices using VST-normalization
#2) Output and use VST-normalized count matrices to generate PCAs
#3) Generate heatmaps with raw count matrices

##NOTE: THIS SCRIPT REQUIRES OUTPUTS FROM "Generate_Consensus_Count_Matrix_Diffbind.R" (since it uses count matrices to generate PCAs)


##################################
rm(list=ls())
library("DESeq2")
library("ggplot2")
library(ChIPseeker)
library("pheatmap")
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
txdb <- TxDb.Mmusculus.UCSC.mm10.knownGene
set.seed(42)


setwd("~/Dropbox/RWY_ATAC_Code_Checking/Code_Checking/")

#First, import the consensus count matrix of raw reads generated by 
consensus_count_matrix <- read.table("Output_Data/FINAL_PEAKSETS_COUNT_MATRICES/consensus_countMatrix_Raw.txt", header=TRUE, sep="\t")
consensus_count_matrix.promoter <- read.table("Output_Data/FINAL_PEAKSETS_COUNT_MATRICES/promoter_consensus_countMatrix_Raw.txt", header=TRUE, sep="\t")

print(colnames(consensus_count_matrix))
#[1] "CHR"      "START"    "END"      "Y_Endo_1" "Y_qNSC_1" "Y_aNSC_1" "Y_NPC_1"  "O_Endo_1" "O_Ast_1"  "O_qNSC_1" "O_aNSC_1" "O_NPC_1"  "Y_Endo_2" "Y_Ast_2"  "Y_qNSC_2" "Y_NPC_2"  "O_Endo_2" "O_Ast_2"  "O_qNSC_2" "O_aNSC_2" "Y_Ast_3"  "Y_qNSC_3" "Y_aNSC_3"
#[24] "Y_NPC_3"  "O_Ast_3"  "O_qNSC_3" "O_aNSC_3" "O_NPC_3" 

###Generate a heatmap of all libraries using Pearson's correlation
pdf("Output_Figs/Consensus_Raw_heatmap_pearson.pdf")
pheatmap(cor(consensus_count_matrix[,c(4:28)], method="pearson"))
dev.off()

#Manually create sampleTable
sampleTable<-  colnames(consensus_count_matrix[,c(4:28)])
sampleTable<- as.data.frame(sampleTable)
names(sampleTable)[1]<- "Sample"
sampleTable$Age <- c(rep(c("Young"),4),rep(c("Old"),5),rep(c("Young"),4),rep(c("Old"),4),rep(c("Young"),4),rep(c("Old"),4))
sampleTable$Cell_Type <- c("Endothelial", "qNSC", "aNSC", "NPC","Endothelial", "Astrocyte", "qNSC", "aNSC", "NPC","Endothelial", "Astrocyte", "qNSC", "NPC","Endothelial", "Astrocyte", "qNSC", "aNSC", "Astrocyte", "qNSC", "aNSC", "NPC", "Astrocyte", "qNSC", "aNSC", "NPC")
sampleTable$Batch <-  c(rep(c("1"),9), rep(c("2"),8), rep(c("3"),8))
sampleTable$Group <- c("Y_Endothelial", "Y_qNSC", "Y_aNSC", "Y_NPC","O_Endothelial", "O_Astrocyte", "O_qNSC", "O_aNSC", "O_NPC","Y_Endothelial", "Y_Astrocyte", "Y_qNSC", "Y_NPC","O_Endothelial", "O_Astrocyte", "O_qNSC", "O_aNSC", "Y_Astrocyte", "Y_qNSC", "Y_aNSC", "Y_NPC", "O_Astrocyte", "O_qNSC", "O_aNSC", "O_NPC")
str(sampleTable)

#Create dds object
dds <- DESeqDataSetFromMatrix(countData=consensus_count_matrix[,c(4:28)], sampleTable, ~Group)
dds.promoter <- DESeqDataSetFromMatrix(countData=consensus_count_matrix.promoter[,c(4:28)], sampleTable, ~Group)

#Calling DESEQ to calculate dispersion estimates
dds <- DESeq(dds)
dds.promoter <- DESeq(dds.promoter)

#Using VST to normalize data
vst <- varianceStabilizingTransformation(dds, blind =F)
vst.promoter <- varianceStabilizingTransformation(dds.promoter, blind =F)

vstMat <- assay(vst)
vstMat.promoter <- assay(vst.promoter)

#Outputting VST-normalized count matrices for downstream analysis (e.g. VST-normalized promoter count matrix is used for correlation with gene expression from Dena's RNA-seq data)
write.table(cbind(consensus_count_matrix[,c(1:3)],as.data.frame(vstMat)), file="Output_Data/FINAL_PEAKSETS_COUNT_MATRICES/consensus_countMatrix_VST.txt",sep="\t", row.names=F,quote=FALSE)
write.table(cbind(consensus_count_matrix.promoter[,c(1:3)],as.data.frame(vstMat.promoter)), file="Output_Data/FINAL_PEAKSETS_COUNT_MATRICES/promoter_consensus_countMatrix_VST.txt",sep="\t", row.names=F,quote=FALSE)



#####################
#VISUALIZATION BY PCA
#####################

#Colors:
#Endo --> "hotpink"
#Ast --> "lightgreen"
#qNSC --> "dodgerblue"
#aNSC --> "red2"
#NPC --> "purple1"

#----------------------------------------
#ALL 25 SAMPLES (CONSENSUS PEAKS)
#----------------------------------------
 
pca.vst <- prcomp(t(vstMat), scale = F)
summary(pca.vst)

#format for ggplot
pca.vst.df <- as.data.frame(pca.vst$x)
str(pca.vst.df)
pca.vst.df$Age <- sampleTable$Age
pca.vst.df$Cell_Type <- sampleTable$Cell_Type
pca.vst.df$Batch <- sampleTable$Batch


theme<-theme(panel.background = element_blank(),panel.border=element_rect(fill=NA),panel.grid.major = element_blank(),panel.grid.minor = element_blank(),strip.background=element_blank(),
             axis.text=element_text(colour="black", size=rel(2)),axis.ticks=element_line(colour="black"),plot.margin=unit(c(1,1,1,1),"line"),
             legend.title = element_text(color = "black", size = 40),legend.text = element_text(color = "black", size = 30))

pdf("Output_Figs/Consensus_PCA.pdf",height=10,width=13)
ggplot(pca.vst.df, aes(PC1,PC2, color=pca.vst.df$Cell_Type ,shape=Age)) +
  geom_point(size=9)+
  scale_color_manual(values=c("hotpink","lightgreen","dodgerblue","red2","purple1"),breaks=c("Endothelial","Astrocyte","qNSC","aNSC","NPC"),name="Cell Type")+
  ggtitle("PCA - all cell types with consensus peakset") +
  xlab("PC1 (26.05%)")+
  ylab("PC2 (14.68%)")+
  theme
dev.off()

#####################################################
#####################################################
#####################################################
#####################################################
#####################################################

#EVERYTHING ABOVE WAS DONE WITH THE GLOBAL CONSENSUS COUNT MATRIX (25 libraries)
#I then load in the Y/O Q/A consensus count matrix (11 libraries) in order to generate a PCA of just Y/O Q/A
#I also load in the count matrices of Y/O Q/A peaks subsetted by genomic element identity (promoter, intron, distal, D+I) to generate PCAs

#----------------------------------------
#ONLY QNSCS AND ANSC (Y+O) SUBSETTED TO GENOMIC ELEMENTS
#----------------------------------------

#COLOR SCHEME:
#Y_qNSC lightskyblue
#O_qNSC dodgerblue
#Y_aNSC orange
#O_aNSC red2

rm(list=ls())
library("DESeq2")
library("ggplot2")
library(ChIPseeker)
library("pheatmap")
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
txdb <- TxDb.Mmusculus.UCSC.mm10.knownGene
set.seed(42)

setwd("~/Dropbox/RWY_ATAC_Code_Checking/Code_Checking/")
consensus_count_matrix <- read.table("Output_Data/FINAL_PEAKSETS_COUNT_MATRICES/Y_O_qNSC_aNSC_countMatrix_Raw.txt", header=TRUE, sep="\t")
consensus_count_matrix.distal <- read.table("Output_Data/FINAL_PEAKSETS_COUNT_MATRICES/distal_Y_O_qNSC_aNSC_countMatrix_Raw.txt", header=TRUE, sep="\t")
consensus_count_matrix.intronic <- read.table("Output_Data/FINAL_PEAKSETS_COUNT_MATRICES/intronic_Y_O_qNSC_aNSC_countMatrix_Raw.txt", header=TRUE, sep="\t")
consensus_count_matrix.promoter <- read.table("Output_Data/FINAL_PEAKSETS_COUNT_MATRICES/promoter_Y_O_qNSC_aNSC_countMatrix_Raw.txt", header=TRUE, sep="\t")
consensus_count_matrix.DI <- read.table("Output_Data/FINAL_PEAKSETS_COUNT_MATRICES/DI_Y_O_qNSC_aNSC_countMatrix_Raw.txt", header=TRUE, sep="\t")

sampleTable<-  colnames(consensus_count_matrix[,c(4:14)])
sampleTable<- as.data.frame(sampleTable)
names(sampleTable)[1]<- "Sample"
sampleTable$Group <- c("Y_qNSC", "Y_aNSC", "O_qNSC","O_aNSC", "Y_qNSC", "O_qNSC","O_aNSC", "Y_qNSC", "Y_aNSC", "O_qNSC","O_aNSC")
str(sampleTable)

dds <- DESeqDataSetFromMatrix(countData=consensus_count_matrix[,c(4:14)], sampleTable, ~Group)
dds.distal <- DESeqDataSetFromMatrix(countData=consensus_count_matrix.distal[,c(4:14)], sampleTable, ~Group)
dds.intronic <- DESeqDataSetFromMatrix(countData=consensus_count_matrix.intronic[,c(4:14)], sampleTable, ~Group)
dds.promoter <- DESeqDataSetFromMatrix(countData=consensus_count_matrix.promoter[,c(4:14)], sampleTable, ~Group)
dds.DI <- DESeqDataSetFromMatrix(countData=consensus_count_matrix.DI[,c(4:14)], sampleTable, ~Group)

dds <- DESeq(dds)
dds.distal <- DESeq(dds.distal)
dds.intronic <- DESeq(dds.intronic)
dds.promoter <- DESeq(dds.promoter)
dds.DI <- DESeq(dds.DI)

vst <- varianceStabilizingTransformation(dds, blind =F)
vst.distal <- varianceStabilizingTransformation(dds.distal, blind =F)
vst.intronic <- varianceStabilizingTransformation(dds.intronic, blind =F)
vst.promoter <- varianceStabilizingTransformation(dds.promoter, blind =F)
vst.DI <- varianceStabilizingTransformation(dds.DI, blind =F)

vstMat <- assay(vst)
vstMat.distal <- assay(vst.distal)
vstMat.intronic <- assay(vst.intronic)
vstMat.promoter <- assay(vst.promoter)
vstMat.DI <- assay(vst.DI)

########################################################
##CONSENSUS

theme<-theme(panel.background = element_blank(),panel.border=element_rect(fill=NA),panel.grid.major = element_blank(),panel.grid.minor = element_blank(),strip.background=element_blank(),
             axis.text=element_text(colour="black", size=rel(2)),axis.ticks=element_line(colour="black"),plot.margin=unit(c(1,1,1,1),"line"),
             legend.title = element_text(color = "black", size = 40),legend.text = element_text(color = "black", size = 30))

pca.vst <- prcomp(t(vstMat), scale = F)
summary(pca.vst)

#Here I am writing out the peaks and associated genes which are driving the principal components
#This will be used for Gene Ontology enrichment and plotted in Supp Fig 4B
PCgenes <- cbind(consensus_count_matrix[,c(1:3)], as.data.frame(pca.vst$rotation))
PC_list <- makeGRangesFromDataFrame(PCgenes, keep.extra.columns = TRUE)
PC_list <- as.data.frame(annotatePeak(PC_list, tssRegion=c(-3000, 3000), TxDb=txdb, annoDb="org.Mm.eg.db"))
write.table(PC_list, sep = "\t", row.names = FALSE, quote = FALSE, file = "Output_Data/Y_O_qNSC_aNSC_PC_list_peaks.txt")

#format for ggplot
pca.vst.df <- as.data.frame(pca.vst$x)
str(pca.vst.df)
pca.vst.df$Age <- c("Young","Young","Old","Old","Young","Old","Old","Young","Young","Old","Old")
pca.vst.df$Cell_Type <- c("qNSC","aNSC","qNSC","aNSC","qNSC","qNSC","aNSC","qNSC","aNSC","qNSC","aNSC")
pca.vst.df$ID <- c("Y_qNSC","Y_aNSC","O_qNSC","O_aNSC","Y_qNSC","O_qNSC","O_aNSC","Y_qNSC","Y_aNSC","O_qNSC","O_aNSC")

#Generating Y_O_Q_A PCA (PC1 vs. PC2)
pdf("Output_Figs/Y_O_Q_A_PC1_PC2.pdf",height=11.8,width=15)
ggplot(pca.vst.df, aes(PC1,PC2, color=pca.vst.df$ID)) +
  geom_point(size=12, shape=18)+
  scale_color_manual(values=c("lightskyblue","orange","dodgerblue","red2"),breaks=c("Y_qNSC","Y_aNSC","O_qNSC","O_aNSC"),labels=c("Young qNSC","Young aNSC","Old qNSC","Old aNSC"), name="")+
  ggtitle("PCA - qNSC and aNSCs") +
  xlab("PC1 (32.58%)")+
  ylab("PC2 (18.27%)")+
  theme
dev.off()

#Generating Y_O_Q_A PCA (PC1 vs. PC3)
pdf("Output_Figs/Y_O_Q_A_PC1_PC3.pdf",height=11.8,width=15)
ggplot(pca.vst.df, aes(PC1,PC3, color=pca.vst.df$ID)) +
  geom_point(size=12, shape=18)+
  scale_color_manual(values=c("lightskyblue","orange","dodgerblue","red2"),breaks=c("Y_qNSC","Y_aNSC","O_qNSC","O_aNSC"),labels=c("Young qNSC","Young aNSC","Old qNSC","Old aNSC"), name="")+
  ggtitle("PCA - qNSC and aNSCs") +
  xlab("PC1 (32.58%)")+
  ylab("PC3 (8.87%)")+
  theme
dev.off()

########################################################
#Supp Figs of D+I, Distal, Intronic, and Promoter PCAs
########################################################

########################################################
##DISTAL + INTRONIC (D+I)

pca.vst <- prcomp(t(vstMat.DI), scale = F)
summary(pca.vst)

#format for ggplot
pca.vst.df <- as.data.frame(pca.vst$x)
str(pca.vst.df)
pca.vst.df$Age <- c("Young","Young","Old","Old","Young","Old","Old","Young","Young","Old","Old")
pca.vst.df$Cell_Type <- c("qNSC","aNSC","qNSC","aNSC","qNSC","qNSC","aNSC","qNSC","aNSC","qNSC","aNSC")
pca.vst.df$ID <- c("Y_qNSC","Y_aNSC","O_qNSC","O_aNSC","Y_qNSC","O_qNSC","O_aNSC","Y_qNSC","Y_aNSC","O_qNSC","O_aNSC")

#Generating Y_O_Q_A PCA based on distal+intronic peaks (PC1 vs. PC3)
pdf("Output_Figs/Y_O_Q_A_PC1_PC3_DI.pdf",height=11.8,width=15)
ggplot(pca.vst.df, aes(PC1,PC3, color=pca.vst.df$ID)) +
  geom_point(size=12, shape=18)+
  scale_color_manual(values=c("lightskyblue","orange","dodgerblue","red2"),breaks=c("Y_qNSC","Y_aNSC","O_qNSC","O_aNSC"),labels=c("Young qNSC","Young aNSC","Old qNSC","Old aNSC"), name="")+
  ggtitle("PCA DI - qNSC and aNSCs") +
  xlab("PC1 (31.75%)")+
  ylab("PC3 (10.04%)")+
  theme
dev.off()

########################################################
##DISTAL

pca.vst <- prcomp(t(vstMat.distal), scale = F)
summary(pca.vst)

#format for ggplot
pca.vst.df <- as.data.frame(pca.vst$x)
str(pca.vst.df)
pca.vst.df$Age <- c("Young","Young","Old","Old","Young","Old","Old","Young","Young","Old","Old")
pca.vst.df$Cell_Type <- c("qNSC","aNSC","qNSC","aNSC","qNSC","qNSC","aNSC","qNSC","aNSC","qNSC","aNSC")
pca.vst.df$ID <- c("Y_qNSC","Y_aNSC","O_qNSC","O_aNSC","Y_qNSC","O_qNSC","O_aNSC","Y_qNSC","Y_aNSC","O_qNSC","O_aNSC")

#Generating Y_O_Q_A PCA based on distal peaks (PC1 vs. PC3)
pdf("Output_Figs/Y_O_Q_A_PC1_PC3_distal.pdf",height=11.8,width=15)
ggplot(pca.vst.df, aes(PC1,PC3, color=pca.vst.df$ID)) +
  geom_point(size=12, shape=18)+
  scale_color_manual(values=c("lightskyblue","orange","dodgerblue","red2"),breaks=c("Y_qNSC","Y_aNSC","O_qNSC","O_aNSC"),labels=c("Young qNSC","Young aNSC","Old qNSC","Old aNSC"), name="")+
  ggtitle("PCA Distal - qNSC and aNSCs") +
  xlab("PC1 (31.55%)")+
  ylab("PC3 (9.95%)")+
  theme + xlim(-45,80) + ylim(-25,55)
dev.off()

########################################################
##INTRONIC

pca.vst <- prcomp(t(vstMat.intronic), scale = F)
summary(pca.vst)

#format for ggplot
pca.vst.df <- as.data.frame(pca.vst$x)
str(pca.vst.df)
pca.vst.df$Age <- c("Young","Young","Old","Old","Young","Old","Old","Young","Young","Old","Old")
pca.vst.df$Cell_Type <- c("qNSC","aNSC","qNSC","aNSC","qNSC","qNSC","aNSC","qNSC","aNSC","qNSC","aNSC")
pca.vst.df$ID <- c("Y_qNSC","Y_aNSC","O_qNSC","O_aNSC","Y_qNSC","O_qNSC","O_aNSC","Y_qNSC","Y_aNSC","O_qNSC","O_aNSC")

#Generating Y_O_Q_A PCA based on intronic peaks (PC1 vs. PC3)
pdf("Output_Figs/Y_O_Q_A_PC1_PC3_intronic.pdf",height=11.8,width=15)
ggplot(pca.vst.df, aes(PC1,PC3, color=pca.vst.df$ID)) +
  geom_point(size=12, shape=18)+
  scale_color_manual(values=c("lightskyblue","orange","dodgerblue","red2"),breaks=c("Y_qNSC","Y_aNSC","O_qNSC","O_aNSC"),labels=c("Young qNSC","Young aNSC","Old qNSC","Old aNSC"), name="")+
  ggtitle("PCA Intronic - qNSC and aNSCs") +
  xlab("PC1 (32.02%)")+
  ylab("PC3 (10.14%)")+
  theme + xlim(-45,80) + ylim(-25,55)
dev.off()

########################################################
##PROMOTER

pca.vst <- prcomp(t(vstMat.promoter), scale = F)
summary(pca.vst)

#format for ggplot
pca.vst.df <- as.data.frame(pca.vst$x)
str(pca.vst.df)
pca.vst.df$Age <- c("Young","Young","Old","Old","Young","Old","Old","Young","Young","Old","Old")
pca.vst.df$Cell_Type <- c("qNSC","aNSC","qNSC","aNSC","qNSC","qNSC","aNSC","qNSC","aNSC","qNSC","aNSC")
pca.vst.df$ID <- c("Y_qNSC","Y_aNSC","O_qNSC","O_aNSC","Y_qNSC","O_qNSC","O_aNSC","Y_qNSC","Y_aNSC","O_qNSC","O_aNSC")

#Generating Y_O_Q_A PCA based on promoter peaks (PC1 vs. PC3)
pdf("Output_Figs/Y_O_Q_A_PC1_PC3_promoter.pdf",height=11.8,width=15)
ggplot(pca.vst.df, aes(PC1,PC3, color=pca.vst.df$ID)) +
  geom_point(size=12, shape=18)+
  scale_color_manual(values=c("lightskyblue","orange","dodgerblue","red2"),breaks=c("Y_qNSC","Y_aNSC","O_qNSC","O_aNSC"),labels=c("Young qNSC","Young aNSC","Old qNSC","Old aNSC"), name="")+
  ggtitle("PCA Promoter - qNSC and aNSCs") +
  xlab("PC1 (36.36%)")+
  ylab("PC3 (10.11%)")+
  theme + xlim(-45,80) + ylim(-25,55)
dev.off()

# RWY sessionInfo()
# R version 3.5.2 (2018-12-20)
# Platform: x86_64-apple-darwin15.6.0 (64-bit)
# Running under: macOS Mojave 10.14.6
# 
# Matrix products: default
# BLAS: /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
# LAPACK: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRlapack.dylib
# 
# locale:
#   [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
# 
# attached base packages:
#   [1] parallel  stats4    stats     graphics  grDevices utils     datasets  methods   base     
# 
# other attached packages:
#   [1] pheatmap_1.0.12                          ggplot2_3.3.2                            DESeq2_1.22.2                            org.Mm.eg.db_3.7.0                       TxDb.Mmusculus.UCSC.mm10.knownGene_3.4.4
# [6] GenomicFeatures_1.34.8                   AnnotationDbi_1.44.0                     ChIPseeker_1.18.0                        DiffBind_2.10.0                          SummarizedExperiment_1.12.0             
# [11] DelayedArray_0.8.0                       BiocParallel_1.16.6                      matrixStats_0.56.0                       Biobase_2.42.0                           GenomicRanges_1.34.0                    
# [16] GenomeInfoDb_1.18.2                      IRanges_2.16.0                           S4Vectors_0.20.1                         BiocGenerics_0.28.0                      data.table_1.12.8                       
# 
# loaded via a namespace (and not attached):
#   [1] backports_1.1.8                         GOstats_2.48.0                          Hmisc_4.3-1                             fastmatch_1.1-0                         plyr_1.8.6                             
# [6] igraph_1.2.5                            GSEABase_1.44.0                         splines_3.5.2                           BatchJobs_1.8                           gridBase_0.4-7                         
# [11] urltools_1.7.3                          amap_0.8-16                             digest_0.6.25                           htmltools_0.5.0                         GOSemSim_2.8.0                         
# [16] viridis_0.5.1                           GO.db_3.7.0                             gdata_2.18.0                            magrittr_1.5                            checkmate_2.0.0                        
# [21] memoise_1.1.0                           BBmisc_1.11                             cluster_2.1.0                           limma_3.38.3                            Biostrings_2.50.2                      
# [26] annotate_1.60.1                         graphlayouts_0.7.0                      systemPipeR_1.16.1                      enrichplot_1.2.0                        prettyunits_1.1.1                      
# [31] colorspace_1.4-1                        blob_1.2.1                              ggrepel_0.8.2                           xfun_0.15                               dplyr_1.0.0                            
# [36] crayon_1.3.4                            RCurl_1.98-1.2                          jsonlite_1.6.1                          TxDb.Hsapiens.UCSC.hg19.knownGene_3.2.2 graph_1.60.0                           
# [41] genefilter_1.64.0                       brew_1.0-6                              survival_3.2-3                          sendmailR_1.2-1                         glue_1.4.1                             
# [46] polyclip_1.10-0                         gtable_0.3.0                            zlibbioc_1.28.0                         XVector_0.22.0                          UpSetR_1.4.0                           
# [51] Rgraphviz_2.26.0                        scales_1.1.1                            DOSE_3.8.2                              DBI_1.1.0                               edgeR_3.24.3                           
# [56] Rcpp_1.0.4.6                            plotrix_3.7-8                           htmlTable_2.0.0                         viridisLite_0.3.0                       xtable_1.8-4                           
# [61] progress_1.2.2                          gridGraphics_0.5-0                      foreign_0.8-76                          bit_1.1-15.2                            europepmc_0.4                          
# [66] Formula_1.2-3                           AnnotationForge_1.24.0                  htmlwidgets_1.5.1                       httr_1.4.1                              fgsea_1.8.0                            
# [71] gplots_3.0.3                            RColorBrewer_1.1-2                      acepack_1.4.1                           ellipsis_0.3.1                          pkgconfig_2.0.3                        
# [76] XML_3.99-0.3                            farver_2.0.3                            nnet_7.3-14                             locfit_1.5-9.4                          ggplotify_0.0.5                        
# [81] tidyselect_1.1.0                        rlang_0.4.6                             reshape2_1.4.4                          munsell_0.5.0                           tools_3.5.2                            
# [86] generics_0.0.2                          RSQLite_2.2.0                           ggridges_0.5.2                          stringr_1.4.0                           yaml_2.2.1                             
# [91] knitr_1.29                              bit64_0.9-7                             tidygraph_1.2.0                         caTools_1.17.1.2                        purrr_0.3.4                            
# [96] ggraph_2.0.3                            RBGL_1.58.2                             DO.db_2.9                               xml2_1.3.2                              biomaRt_2.38.0                         
# [101] compiler_3.5.2                          rstudioapi_0.11                         geneplotter_1.60.0                      tibble_3.0.1                            tweenr_1.0.1                           
# [106] stringi_1.4.6                           lattice_0.20-41                         Matrix_1.2-18                           vctrs_0.3.1                             pillar_1.4.4                           
# [111] lifecycle_0.2.0                         BiocManager_1.30.10                     triebeard_0.3.0                         cowplot_1.0.0                           bitops_1.0-6                           
# [116] rtracklayer_1.42.2                      qvalue_2.14.1                           R6_2.4.1                                latticeExtra_0.6-28                     hwriter_1.3.2                          
# [121] ShortRead_1.40.0                        KernSmooth_2.23-16                      gridExtra_2.3                           boot_1.3-25                             MASS_7.3-51.6                          
# [126] gtools_3.8.2                            Category_2.48.1                         rjson_0.2.20                            withr_2.2.0                             GenomicAlignments_1.18.1               
# [131] Rsamtools_1.34.1                        GenomeInfoDbData_1.2.0                  hms_0.5.3                               rpart_4.1-15                            grid_3.5.2                             
# [136] tidyr_1.1.0                             rvcheck_0.1.8                           ggforce_0.3.2                           base64enc_0.1-3                        



