##################################
#04/02/2020

#VST Normalization --> Plotting PCAs

#Purpose:
#1) Normalize raw consensus count matrix using VST-normalization
#2) Output and use VST-normalized count matrices to generate PCAs
#3) Generate heatmaps with raw count matrices


##################################
rm(list=ls())
library("DESeq2")
library("ggplot2")
library(ChIPseeker)
library("pheatmap")
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
txdb <- TxDb.Mmusculus.UCSC.mm10.knownGene
set.seed(42)



setwd("~/Dropbox/RWY_ATAC_Code_Checking/Code_Checking/")

#First, import the consensus count matrix of raw reads generated by 
consensus_count_matrix <- read.table("Original_Data/Leeman_RNAseq_expression_values.txt", header=TRUE, row.names = 1, sep="\t")


print(colnames(consensus_count_matrix))
#[1] "CHR"      "START"    "END"      "Y_Endo_1" "Y_qNSC_1" "Y_aNSC_1" "Y_NPC_1"  "O_Endo_1" "O_Ast_1"  "O_qNSC_1" "O_aNSC_1" "O_NPC_1"  "Y_Endo_2" "Y_Ast_2"  "Y_qNSC_2" "Y_NPC_2"  "O_Endo_2" "O_Ast_2"  "O_qNSC_2" "O_aNSC_2" "Y_Ast_3"  "Y_qNSC_3" "Y_aNSC_3"
#[24] "Y_NPC_3"  "O_Ast_3"  "O_qNSC_3" "O_aNSC_3" "O_NPC_3" 



###Generate a heatmap of all libraries using Pearson's correlation
pdf("Output_Figs/Consensus_Raw_heatmap_pearson.pdf")
pheatmap(cor(consensus_count_matrix, method="spearman"))
dev.off()


#Manually create sampleTable
sampleTable<-  colnames(consensus_count_matrix)
sampleTable<- as.data.frame(sampleTable)
names(sampleTable)[1]<- "Sample"
sampleTable$Age <- c(rep(c("Young"),4),rep(c("Old"),4),rep(c("Young"),2),rep(c("Old"),3),rep(c("Young"),3),rep(c("Old"),3),rep(c("Young"),4),rep(c("Old"),4),rep(c("Young"),4),rep(c("Old"),4))
sampleTable$Cell_Type <- c(rep(c("Astrocyte"),8),rep(c("qNSC"),5),rep(c("aNSC"),6),rep(c("NPC"),8),rep(c("Endo"),8))
sampleTable$Group <- c(rep(c("Y_Ast"),4),rep(c("O_Ast"),4),rep(c("Y_qNSC"),2),rep(c("O_qNSC"),3),rep(c("Y_aNSC"),3),rep(c("O_aNSC"),3),rep(c("Y_NPC"),4),rep(c("O_NPC"),4),rep(c("Y_Endo"),4),rep(c("O_Endo"),4))
str(sampleTable)



#----------------------------------------
#ALL 25 SAMPLES (CONSENSUS PEAKS)
#----------------------------------------
 
pca.vst <- prcomp(t(consensus_count_matrix), scale = F)
summary(pca.vst)

#format for ggplot
pca.vst.df <- as.data.frame(pca.vst$x)
str(pca.vst.df)
pca.vst.df$Age <- sampleTable$Age
pca.vst.df$Cell_Type <- sampleTable$Cell_Type


theme<-theme(panel.background = element_blank(),panel.border=element_rect(fill=NA),panel.grid.major = element_blank(),panel.grid.minor = element_blank(),strip.background=element_blank(),
             axis.text=element_text(colour="black", size=rel(2)),axis.ticks=element_line(colour="black"),plot.margin=unit(c(1,1,1,1),"line"),
             legend.title = element_text(color = "black", size = 20),legend.text = element_text(color = "black", size = 16))

pdf("Output_Figs/Consensus_PCA.pdf",height=10,width=12)
ggplot(pca.vst.df, aes(PC1,PC2, color=Cell_Type ,shape=Age)) +
  geom_point(size=9)+
  scale_color_manual(values=c("lightgreen","mediumturquoise","royalblue1","purple1","orange1"),breaks=c("Astrocyte","qNSC","aNSC","NPC","Endothelial"),name="Cell Type")+
  ggtitle("PCA - all cell types with consensus peakset") +
  xlab("PC1 (26.05%)")+
  ylab("PC2 (14.68%)")+
  theme
dev.off()




#####################################################
#####################################################
#####################################################
#####################################################
#####################################################

#EVERYTHING ABOVE WAS DONE WITH THE GLOBAL CONSENSUS PEAKSET
#I then load in the young+old qNSC+aNSC consensus peakset in order to generate a PCA of just Y/O Q/A


#----------------------------------------
#ONLY QNSCS AND ANSC (Y+O)
#----------------------------------------

NSC_matrix <- consensus_count_matrix[,c(9:19)]

sampleTable<-  colnames(NSC_matrix)
sampleTable<- as.data.frame(sampleTable)
names(sampleTable)[1]<- "Sample"
sampleTable$Group <- c(rep(c("Y_qNSC"),2),rep(c("O_qNSC"),3),rep(c("Y_aNSC"),3),rep(c("O_aNSC"),3))
str(sampleTable)

pca.vst <- prcomp(t(NSC_matrix), scale = F)
summary(pca.vst)

pca.vst.df <- as.data.frame(pca.vst$x)
str(pca.vst.df)
pca.vst.df$Age <- c(rep(c("Young","Young","Old","Old","Old","Young","Young","Young","Old","Old","Old")))
pca.vst.df$Cell_Type <- c(rep(c("qNSC"),5),rep(c("aNSC"),6))



theme<-theme(panel.background = element_blank(),panel.border=element_rect(fill=NA),panel.grid.major = element_blank(),panel.grid.minor = element_blank(),strip.background=element_blank(),
             axis.text=element_text(colour="black", size=rel(2)),axis.ticks=element_line(colour="black"),plot.margin=unit(c(1,1,1,1),"line"),
             legend.title = element_text(color = "black", size = 20),legend.text = element_text(color = "black", size = 16))

pdf("Output_Figs/RNA_Y_O_Q_A_PC1_PC2.pdf",height=10.5,width=12)
ggplot(pca.vst.df, aes(PC1,PC2, color=pca.vst.df$Cell_Type ,shape=Age)) +
  geom_point(size=9)+
  scale_color_manual(values=c("dodgerblue","red2"),breaks=c("qNSC","aNSC"),name="Cell Type")+
  ggtitle("PCA - qNSC and aNSCs") +
  xlab("PC1 (58.18%)")+
  ylab("PC2 (9.96%)")+
  theme
dev.off()

pdf("Output_Figs/RNA_Y_O_Q_A_PC1_PC3.pdf",height=10.5,width=12)
ggplot(pca.vst.df, aes(PC1,PC3, color=pca.vst.df$Cell_Type ,shape=Age)) +
  geom_point(size=9)+
  scale_color_manual(values=c("dodgerblue","red2"),breaks=c("qNSC","aNSC"),name="Cell Type")+
  ggtitle("PCA - qNSC and aNSCs") +
  xlab("PC1 (58.18%)")+
  ylab("PC3 (6.25%)")+
  theme
dev.off()

pdf("Output_Figs/RNA_Y_O_Q_A_PC1_PC4.pdf",height=10.5,width=12)
ggplot(pca.vst.df, aes(PC1,PC4, color=pca.vst.df$Cell_Type ,shape=Age)) +
  geom_point(size=9)+
  scale_color_manual(values=c("dodgerblue","red2"),breaks=c("qNSC","aNSC"),name="Cell Type")+
  ggtitle("PCA - qNSC and aNSCs") +
  xlab("PC1 (58.18%)")+
  ylab("PC3 (5.77%)")+
  theme
dev.off()




#----------------------------------------
#ONLY QNSCS AND ANSC (Y+O) SUBSETTED TO GENOMIC ELEMENTS
#----------------------------------------

rm(list=ls())
library("DESeq2")
library("ggplot2")
library(ChIPseeker)
library("pheatmap")
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
txdb <- TxDb.Mmusculus.UCSC.mm10.knownGene
set.seed(42)

consensus_count_matrix <- read.table("Output_Data/FINAL_PEAKSETS_COUNT_MATRICES/Y_O_qNSC_aNSC_countMatrix_Raw.txt", header=TRUE, sep="\t")
consensus_count_matrix.distal <- read.table("Output_Data/FINAL_PEAKSETS_COUNT_MATRICES/distal_Y_O_qNSC_aNSC_countMatrix_Raw.txt", header=TRUE, sep="\t")
consensus_count_matrix.intronic <- read.table("Output_Data/FINAL_PEAKSETS_COUNT_MATRICES/intronic_Y_O_qNSC_aNSC_countMatrix_Raw.txt", header=TRUE, sep="\t")
consensus_count_matrix.promoter <- read.table("Output_Data/FINAL_PEAKSETS_COUNT_MATRICES/promoter_Y_O_qNSC_aNSC_countMatrix_Raw.txt", header=TRUE, sep="\t")
consensus_count_matrix.DI <- read.table("Output_Data/FINAL_PEAKSETS_COUNT_MATRICES/DI_Y_O_qNSC_aNSC_countMatrix_Raw.txt", header=TRUE, sep="\t")

sampleTable<-  colnames(consensus_count_matrix[,c(4:14)])
sampleTable<- as.data.frame(sampleTable)
names(sampleTable)[1]<- "Sample"
sampleTable$Group <- c("Y_qNSC", "Y_aNSC", "O_qNSC","O_aNSC", "Y_qNSC", "O_qNSC","O_aNSC", "Y_qNSC", "Y_aNSC", "O_qNSC","O_aNSC")
str(sampleTable)

dds <- DESeqDataSetFromMatrix(countData=consensus_count_matrix[,c(4:14)], sampleTable, ~Group)
dds.distal <- DESeqDataSetFromMatrix(countData=consensus_count_matrix.distal[,c(4:14)], sampleTable, ~Group)
dds.intronic <- DESeqDataSetFromMatrix(countData=consensus_count_matrix.intronic[,c(4:14)], sampleTable, ~Group)
dds.promoter <- DESeqDataSetFromMatrix(countData=consensus_count_matrix.promoter[,c(4:14)], sampleTable, ~Group)
dds.DI <- DESeqDataSetFromMatrix(countData=consensus_count_matrix.DI[,c(4:14)], sampleTable, ~Group)

dds <- DESeq(dds)
dds.distal <- DESeq(dds.distal)
dds.intronic <- DESeq(dds.intronic)
dds.promoter <- DESeq(dds.promoter)
dds.DI <- DESeq(dds.DI)

vst <- varianceStabilizingTransformation(dds, blind =F)
vst.distal <- varianceStabilizingTransformation(dds.distal, blind =F)
vst.intronic <- varianceStabilizingTransformation(dds.intronic, blind =F)
vst.promoter <- varianceStabilizingTransformation(dds.promoter, blind =F)
vst.DI <- varianceStabilizingTransformation(dds.DI, blind =F)

vstMat <- assay(vst)
vstMat.distal <- assay(vst.distal)
vstMat.intronic <- assay(vst.intronic)
vstMat.promoter <- assay(vst.promoter)
vstMat.DI <- assay(vst.DI)


########################################################
##CONSENSUS

pca.vst <- prcomp(t(vstMat), scale = F)
summary(pca.vst)

#format for ggplot
pca.vst.df <- as.data.frame(pca.vst$x)
str(pca.vst.df)
pca.vst.df$Age <- c(rep(c("Young","Young","Old","Old","Young","Old","Old","Young","Young","Old","Old")))
pca.vst.df$Cell_Type <- c("qNSC","aNSC","qNSC","aNSC","qNSC","qNSC","aNSC","qNSC","aNSC","qNSC","aNSC")

theme<-theme(panel.background = element_blank(),panel.border=element_rect(fill=NA),panel.grid.major = element_blank(),panel.grid.minor = element_blank(),strip.background=element_blank(),
             axis.text=element_text(colour="black", size=rel(2)),axis.ticks=element_line(colour="black"),plot.margin=unit(c(1,1,1,1),"line"),
             legend.title = element_text(color = "black", size = 20),legend.text = element_text(color = "black", size = 16))

pdf("Output_Figs/Y_O_Q_A_PC1_PC3_new.pdf",height=10.5,width=12)
ggplot(pca.vst.df, aes(PC1,PC3, color=pca.vst.df$Cell_Type ,shape=Age)) +
  geom_point(size=9)+
  scale_color_manual(values=c("dodgerblue","red2"),breaks=c("qNSC","aNSC"),name="Cell Type")+
  ggtitle("PCA - qNSC and aNSCs") +
  xlab("PC1 (32.58%)")+
  ylab("PC3 (8.87%)")+
  theme + xlim(-45,80) + ylim(-25,55)
dev.off()


########################################################
##DISTAL

pca.vst <- prcomp(t(vstMat.distal), scale = F)
summary(pca.vst)

#format for ggplot
pca.vst.df <- as.data.frame(pca.vst$x)
str(pca.vst.df)
pca.vst.df$Age <- c(rep(c("Young","Young","Old","Old","Young","Old","Old","Young","Young","Old","Old")))
pca.vst.df$Cell_Type <- c("qNSC","aNSC","qNSC","aNSC","qNSC","qNSC","aNSC","qNSC","aNSC","qNSC","aNSC")

theme<-theme(panel.background = element_blank(),panel.border=element_rect(fill=NA),panel.grid.major = element_blank(),panel.grid.minor = element_blank(),strip.background=element_blank(),
             axis.text=element_text(colour="black", size=rel(2)),axis.ticks=element_line(colour="black"),plot.margin=unit(c(1,1,1,1),"line"),
             legend.title = element_text(color = "black", size = 20),legend.text = element_text(color = "black", size = 16))

pdf("Output_Figs/Y_O_Q_A_PC1_PC3_distal_new.pdf",height=10.5,width=12)
ggplot(pca.vst.df, aes(PC1,PC3, color=pca.vst.df$Cell_Type ,shape=Age)) +
  geom_point(size=9)+
  scale_color_manual(values=c("dodgerblue","red2"),breaks=c("qNSC","aNSC"),name="Cell Type")+
  ggtitle("PCA Distal - qNSC and aNSCs") +
  xlab("PC1 (31.55%)")+
  ylab("PC3 (9.95%)")+
  theme + xlim(-45,80) + ylim(-25,55)
dev.off()


########################################################
##INTRONIC

pca.vst <- prcomp(t(vstMat.intronic), scale = F)
summary(pca.vst)

#format for ggplot
pca.vst.df <- as.data.frame(pca.vst$x)
str(pca.vst.df)
pca.vst.df$Age <- c(rep(c("Young","Young","Old","Old","Young","Old","Old","Young","Young","Old","Old")))
pca.vst.df$Cell_Type <- c("qNSC","aNSC","qNSC","aNSC","qNSC","qNSC","aNSC","qNSC","aNSC","qNSC","aNSC")

theme<-theme(panel.background = element_blank(),panel.border=element_rect(fill=NA),panel.grid.major = element_blank(),panel.grid.minor = element_blank(),strip.background=element_blank(),
             axis.text=element_text(colour="black", size=rel(2)),axis.ticks=element_line(colour="black"),plot.margin=unit(c(1,1,1,1),"line"),
             legend.title = element_text(color = "black", size = 20),legend.text = element_text(color = "black", size = 16))

pdf("Output_Figs/Y_O_Q_A_PC1_PC3_intronic_new.pdf",height=10.5,width=12)
ggplot(pca.vst.df, aes(PC1,PC3, color=pca.vst.df$Cell_Type ,shape=Age)) +
  geom_point(size=9)+
  scale_color_manual(values=c("dodgerblue","red2"),breaks=c("qNSC","aNSC"),name="Cell Type")+
  ggtitle("PCA Intronic - qNSC and aNSCs") +
  xlab("PC1 (32.02%)")+
  ylab("PC3 (10.14%)")+
  theme + xlim(-45,80) + ylim(-25,55)
dev.off()




########################################################
##PROMOTER

pca.vst <- prcomp(t(vstMat.promoter), scale = F)
summary(pca.vst)

#format for ggplot
pca.vst.df <- as.data.frame(pca.vst$x)
str(pca.vst.df)
pca.vst.df$Age <- c(rep(c("Young","Young","Old","Old","Young","Old","Old","Young","Young","Old","Old")))
pca.vst.df$Cell_Type <- c("qNSC","aNSC","qNSC","aNSC","qNSC","qNSC","aNSC","qNSC","aNSC","qNSC","aNSC")

theme<-theme(panel.background = element_blank(),panel.border=element_rect(fill=NA),panel.grid.major = element_blank(),panel.grid.minor = element_blank(),strip.background=element_blank(),
             axis.text=element_text(colour="black", size=rel(2)),axis.ticks=element_line(colour="black"),plot.margin=unit(c(1,1,1,1),"line"),
             legend.title = element_text(color = "black", size = 20),legend.text = element_text(color = "black", size = 16))

pdf("Output_Figs/Y_O_Q_A_PC1_PC3_promoter_new.pdf",height=10.5,width=12)
ggplot(pca.vst.df, aes(PC1,PC3, color=pca.vst.df$Cell_Type ,shape=Age)) +
  geom_point(size=9)+
  scale_color_manual(values=c("dodgerblue","red2"),breaks=c("qNSC","aNSC"),name="Cell Type")+
  ggtitle("PCA Promoter - qNSC and aNSCs") +
  xlab("PC1 (36.36%)")+
  ylab("PC3 (10.11%)")+
  theme + xlim(-45,80) + ylim(-25,55)
dev.off()


########################################################
##DISTAL + INTRONIC

pca.vst <- prcomp(t(vstMat.DI), scale = F)
summary(pca.vst)

#format for ggplot
pca.vst.df <- as.data.frame(pca.vst$x)
str(pca.vst.df)
pca.vst.df$Age <- c(rep(c("Young","Young","Old","Old","Young","Old","Old","Young","Young","Old","Old")))
pca.vst.df$Cell_Type <- c("qNSC","aNSC","qNSC","aNSC","qNSC","qNSC","aNSC","qNSC","aNSC","qNSC","aNSC")

theme<-theme(panel.background = element_blank(),panel.border=element_rect(fill=NA),panel.grid.major = element_blank(),panel.grid.minor = element_blank(),strip.background=element_blank(),
             axis.text=element_text(colour="black", size=rel(2)),axis.ticks=element_line(colour="black"),plot.margin=unit(c(1,1,1,1),"line"),
             legend.title = element_text(color = "black", size = 20),legend.text = element_text(color = "black", size = 16))

pdf("Output_Figs/Y_O_Q_A_PC1_PC3_DI_new.pdf",height=10.5,width=12)
ggplot(pca.vst.df, aes(PC1,PC3, color=pca.vst.df$Cell_Type ,shape=Age)) +
  geom_point(size=9)+
  scale_color_manual(values=c("dodgerblue","red2"),breaks=c("qNSC","aNSC"),name="Cell Type")+
  ggtitle("PCA DI - qNSC and aNSCs") +
  xlab("PC1 (31.75%)")+
  ylab("PC3 (10.04%)")+
  theme + xlim(-45,80) + ylim(-25,55)
dev.off()























